File: leetcode_submissions/2024-10-04/234. Palindrome Linked List/21. Merge Two Sorted Lists/1256. Rank Transform of an Array/1620. Check If Array Pairs Are Divisible by k/1620. Check If Array Pairs Are Divisible by k/Check_If_Array_Pairs_Are_Divisible_by_k.py
class Solution:
    def canArrange(self, arr: List[int], k: int) -> bool:
        
        n = len(arr)

        if n%2 !=0:
            return 0

        remainder_count = defaultdict(int)

        for num in arr:
            remainder = (num%k + k) %k

            remainder_count[remainder] += 1


        for i in range(k //2):
            if i == 0:
                if remainder_count[0]%2 != 0:
                    return 0

            else:
                if remainder_count[i] != remainder_count[k-i]:
                    return 0

        return 1

File: leetcode_submissions/2024-10-03/1620_Check_If_Array_Pairs_Are_Divisible_by_k/Check_If_Array_Pairs_Are_Divisible_by_k.cpp
class Solution {
public:
    bool canArrange(vector<int>& arr, int k) {
        int n = arr.size();
        
        // Ensure that the array has an even number of elements
        if (n % 2 != 0) return false;
        
        // Pre-process remainders
        for (int i = 0; i < n; ++i) {
            // Calculate remainder and handle negative numbers correctly
            arr[i] = ((arr[i] % k) + k) % k;
        }

        // Iterate over each element to try and find a pair
        for (int i = 0; i < n; ++i) {
            // Skip if this element has already been paired (marked as -1)
            if (arr[i] == -1) continue;

            bool foundPair = false; // Flag to check if a pair was found
            
            // Try to find a pair for arr[i]
            for (int j = i + 1; j < n; ++j) {
                // Skip if this element has already been paired
                if (arr[j] == -1) continue;
                
                // Check if the sum of the remainders of arr[i] and arr[j] is divisible by k
                if ((arr[i] + arr[j]) % k == 0) {
                    // Mark both elements as paired by setting them to -1
                    arr[i] = -1;
                    arr[j] = -1;
                    foundPair = true;
                    break; // Exit the inner loop since we found a pair for arr[i]
                }
            }
            
            // If no pair was found for arr[i], return false
            if (!foundPair) return false;
        }
        
        // If all elements have been successfully paired, return true
        return true;
    }
};
